### PR TITLE
Add Beckon Dfu

### PR DESCRIPTION
**Pull request by Rawa**

### CODE REVIEW

File: beckon-dfu/.gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+/build
```
Comment (lenguyenthanh) [inline beckon-dfu/.gitignore:1]: wtf

File: beckon-dfu/src/main/java/com/technocreatives/beckon/dfu/DfuClient.kt [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+sealed interface StartDfuError {
+    object AlreadyInProgress : StartDfuError
+}
+
+data class DfuConfiguration(
+    val macAddress: String,
+    val firmware: Int, // Todo change to Uri
+    val dfuService: Class<out DfuBaseService>,
```
Comment (lenguyenthanh) [inline beckon-dfu/src/main/java/com/technocreatives/beckon/dfu/DfuClient.kt:32]: should we hide this from users?

File: beckon-dfu/src/main/java/com/technocreatives/beckon/dfu/DfuProcess.kt [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    suspend fun abort(): Either<AbortError, Unit>
+}
+
+sealed interface AbortError {
+    object AlreadyAborted : AbortError
+    object Timeout : AbortError
+}
+
+sealed interface DfuState {
+    fun finished(): Boolean = when (this) {
```
Comment (lenguyenthanh) [inline beckon-dfu/src/main/java/com/technocreatives/beckon/dfu/DfuProcess.kt:20]: Maybe `isStop` a better name

File: beckon-dfu/src/main/java/com/technocreatives/beckon/dfu/internal/DfuProcessImpl.kt [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+        scope.launch {
+            dfuEvents.scan(initialState, reducer)
+                .collect(_mutableDfuState)
+            Timber.d("Dfu state collection is complete")
+        }
+    }
+
+    override suspend fun abort(): Either<AbortError, Unit> {
+        if (dfuServiceController.isAborted) {
```
Comment (lenguyenthanh) [inline beckon-dfu/src/main/java/com/technocreatives/beckon/dfu/internal/DfuProcessImpl.kt:53]: need to check if the process is finished as well.

File: beckon-dfu/src/main/java/com/technocreatives/beckon/dfu/DfuClient.kt (hunk 2 above)
Comment (Rawa) [reply beckon-dfu/src/main/java/com/technocreatives/beckon/dfu/DfuClient.kt]: 😢 No

**Pull request merged**